src/cargo/core/compiler/job_queue.rs

```diff
@@ -251,26 +384,203 @@ impl<'a, 'cfg> JobQueue<'a, 'cfg> {
+                    }
+                };
+                info!("end ({:?}): {:?}", unit, result);
+                match result {
+                    Ok(()) => self.finish(id, &unit, artifact, cx)?,
+                    Err(e) => {
+                        let msg = "The following warnings were emitted during compilation:";
+                        self.emit_warnings(Some(msg), &unit, cx)?;
+
+                        if !self.active.is_empty() {
+                            handle_error(&e, &mut *cx.bcx.config.shell());
+                            cx.bcx.config.shell().warn(
+                                "build failed, waiting for other \
+                                 jobs to finish...",
+                            )?;
+                            return Ok(Some(anyhow::format_err!("build failed")));
+                        } else {
+                            return Ok(Some(e));
+                        }
+                    }
+                }
+            }
+            Message::Token(acquired_token) => {
+                let token = acquired_token.chain_err(|| "failed to acquire jobserver token")?;
+                self.tokens.push(token);
+            }
+            Message::NeedsToken(id, client) => {
+                log::info!("queue token request");
+                jobserver_helper.request_token();
+                self.to_send_clients.push((id, client));
+            }
+            Message::ReleaseToken(id) => {
+                // Note that this pops off potentially a completely
+                // different token, but all tokens of the same job are
+                // conceptually the same so that's fine.
+                if let Some(rustc_tokens) = self.rustc_tokens.get_mut(&id) {
+                    self.tokens
+                        .push(rustc_tokens.pop().expect("rustc releases token it has"));
+                } else {
+                    panic!("This job does not have tokens associated with it");
+                }
+            }
+        }
+
+        Ok(None)
+    }
+
+    // This will also tick the progress bar as appropriate
+    fn wait_for_events(&mut self) -> Vec<Message> {
+        // Drain all events at once to avoid displaying the progress bar
+        // unnecessarily. If there's no events we actually block waiting for
+        // an event, but we keep a "heartbeat" going to allow `record_cpu`
+        // to run above to calculate CPU usage over time. To do this we
+        // listen for a message with a timeout, and on timeout we run the
+        // previous parts of the loop again.
+        let events: Vec<_> = self.rx.try_iter().collect();
+        if events.is_empty() {
+            loop {
+                self.tick_progress();
+                match self.rx.recv_timeout(Duration::from_millis(500)) {
+                    Ok(message) => break vec![message],
+                    Err(_) => continue,
+                }
+            }
+        } else {
+            self.tick_progress();
```

Comment:
I think the purpose of this loop is to actually not call this in the case that the events are not empty, so was this perhaps a stray refactoring error?

Reply:
Especially in the case where rustc is using this channel for the jobserver protocol, events may not be empty for a long time (i.e., we're constantly trying to wait for threads, releasing tokens, etc.). I can drop this from this PR though and file it separately since it's a drive by change if anything.

Reply:
Hm ok, mind moving this to a separate commit? I think it'd be best to handle this with a sort of heartbeat message if it becomes problematic. Did you see extraneous stuttering you weren't expecting when building though?